File: 3652-best-time-to-buy-and-sell-stock-using-strategy/3652-best-time-to-buy-and-sell-stock-using-strategy.cpp
class Solution {
public:
    long long maxProfit(vector<int>& prices, vector<int>& strategy, int k) {
        long long res = 0;
        long long sum = 0;
        long long windowSum = 0;
        int l = 0;
        vector<int> pre(prices.size(), 0);
        for(int i = 0; i < prices.size(); i++)
        {
            sum += (prices[i] * strategy[i]);
            windowSum += (prices[i] * strategy[i]);
            if(i >= k - 1)
            {
                pre[i] = windowSum;
                windowSum -= (prices[l] * strategy[l]);
                l++;
            }
        }
        res = sum;
        long long profit = 0;
        l = 0;
        for(int r = 0; r < prices.size(); r++)
        {
            if(r >= k / 2)
            {
                profit += prices[r];
            }
            if(r - l + 1 == k)
            {
                res = max(res, sum - pre[r] + profit);
                profit -= prices[l + (k / 2)];
                l++;
            }
        }
        return res;
    }
};

File: 3652-best-time-to-buy-and-sell-stock-using-strategy/3652-best-time-to-buy-and-sell-stock-using-strategy.py
class Solution:
    def maxProfit(self, prices: List[int], strategy: List[int], k: int) -> int:
        res = 0
        tot = 0
        windowTot = 0
        profit = 0
        l = 0
        pre = [0 for _ in range(len(prices))]
        for r in range(len(prices)):
            tot += (prices[r] * strategy[r])
            windowTot += (prices[r] * strategy[r])
            if r >= k - 1:
                pre[r] = windowTot
                windowTot -= (prices[l] * strategy[l])
                l += 1
        res = tot
        l = 0
        for r in range(len(prices)):
            if r >= (k / 2):
                profit += prices[r]
            if r - l + 1 == k:
                res = max(res, tot - pre[r] + profit)
                profit -= prices[l + (k // 2)]
                l += 1
        return res
